Introduce MemoryDump printing utility for dataLog()

ast-hugger opened the pull request:
#### fdc2659799c04e8f04466f8726c3a73b198b13e1
<pre>
Introduce MemoryDump printing utility for dataLog()
<a href="https://bugs.webkit.org/show_bug.cgi?id=304214">https://bugs.webkit.org/show_bug.cgi?id=304214</a>
<a href="https://rdar.apple.com/166570876">rdar://166570876</a>

Reviewed by Dan Hecht and Keith Miller.

It&apos;s sometimes helpful in trace output to print relatively large chunks of raw data such as
stack frames in the traditional hex dump form. This patch introduces a new printing
plug-in for dataLog() to do just that.

Examples:

    dataLogLn(&quot;Interesting data:&quot;, MemoryDump(startPtr, endPtr));
    dataLogLn(&quot;Interesting data:&quot;, MemoryDump(std::span(data)));

The style checker is unhappy with the usage of &apos;out.printf()&apos; in the code added to
PrintStream.cpp, however that is consistent with the rest of that file.

Tests: Tools/TestWebKitAPI/Tests/WTF/MemoryDump.cpp
Canonical link: <a href="https://commits.webkit.org/304560@main">https://commits.webkit.org/304560@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/73a90b80e5241941e7ee9cc498545d35958fcf08

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135891 "12 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8266 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47182 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143604 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/87459 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d37f1c2c-0e3a-4258-93c4-ca156caa49c2) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137760 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8924 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8111 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103862 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/87459 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/bb40f8fa-acbe-4bcb-8f5d-66b76f604a92) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138837 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6462 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121803 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84738 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6167 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3806 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4207 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/127869 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/115431 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39991 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146352 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/134375 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7950 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40557 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112216 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7973 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6677 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112604 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28581 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6076 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118101 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/61851 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7997 "Built successfully") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36182 "Found 1 new test failure: fast/mediastream/granted-denied-request-management2.html (failure)") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/167181 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7727 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71549 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43635 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7948 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7810 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/WTF/wtf/HexDump.h:
```diff
@@ -0,0 +1,64 @@
+
+private:
+    const void* m_data { nullptr };
+    size_t m_size { 0 };
```
Why not just make this a std::span? I think you'll get a SaferCPP warning about this anyway.

ast-hugger replied:
SaferCPP can't detect something like this, but a span makes sense. However, I'm intentionally keeping a constructor that takes a pointer pair, for ergonomics. The compiler is picky about spans and wants unsafe usage pragmas to create one over an arbitrary chunk of memory, but for debugging usage this is often exactly the use case. You have two pointers and want to see the memory in between.

kmiller68 commented on Source/WTF/wtf/HexDump.h:
```diff
@@ -0,0 +1,64 @@
+//   dataLogLn("Array dump: ", HexDump(std::span(array)));
+//   dataLogLn("Vector dump: ", HexDump(std::span(vector)));
+
+class HexDump {
```
I think HexSpan would be a clearer name to distinguish this from RawHex.

ast-hugger replied:
I intentionally used "dump" to emphasize that it's more "imperative" than other constructs you use with print streams, which are basically just data transformers whose output is embedded into what's usually a simple line of text. I would imagine HexSpan as a formatter that produces a simple sequence of byte values in the span, like "41 32 7f ab". Compared to that, a memory dump is a different presentation mode that produces a whole self-contained table of memory addresses and values.

kmiller68 replied:
Hmm, there are lots of classes with "Dump" in them that are largely just formatters meant to be embedded inline with other text. e.g. PointerDump, EnumDump, DumpInContext, etc.

What about something like HexSpanDump or HexBulkDump?

ast-hugger replied:
But what's wrong with just HexDump? This is what exactly this kind of output has been called for decades. There is a `hexdump` utility in macOS. What extra information would "span" or "bulk" add?

ast-hugger replied:
Though I also like MemoryDump. Short and to the point.

dhecht commented on Source/WTF/wtf/MemoryDump.h:
```diff
@@ -0,0 +1,75 @@
+        , m_sizeLimit(sizeLimit)
+    { }
+
+    explicit MemoryDump(const void* ptr1, const void* ptr2, size_t sizeLimit = DefaultSizeLimit);
```
maybe call them `ptrStart` and `ptrEnd` to be clearer that the second pointer is not included in the span

ast-hugger replied:
But that's not quite how they work. They are compared, and the lesser is used as the start and the greater as the end not included in the span.

dhecht reviewed (approved):
r=me with nit

kmiller68 commented on Source/WTF/wtf/MemoryDump.h:
```diff
@@ -0,0 +1,75 @@
+
+private:
+    std::span<const std::byte> m_data { };
+    size_t m_sizeLimit { DefaultSizeLimit };
```
Might be worth a FIXME: to support different hunk sizes e.g. 16/32/64 bits.

kmiller68 reviewed (approved):
r=me too.

ast-hugger committed:
Introduce MemoryDump printing utility for dataLog()
https://bugs.webkit.org/show_bug.cgi?id=304214
rdar://166570876

Reviewed by Dan Hecht and Keith Miller.

It's sometimes helpful in trace output to print relatively large chunks of raw data such as
stack frames in the traditional hex dump form. This patch introduces a new printing
plug-in for dataLog() to do just that.

Examples:

    dataLogLn("Interesting data:", MemoryDump(startPtr, endPtr));
    dataLogLn("Interesting data:", MemoryDump(std::span(data)));

The style checker is unhappy with the usage of 'out.printf()' in the code added to
PrintStream.cpp, however that is consistent with the rest of that file.

Tests: Tools/TestWebKitAPI/Tests/WTF/MemoryDump.cpp
Canonical link: https://commits.webkit.org/304560@main
